### PR TITLE
[apps] Enable Clang-Format for in_hand_scanner

### PR DESCRIPTION
**Pull request by SunBlack**

### CODE REVIEW

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/common_types.h [hunk 1]
```diff
@@ -40,55 +40,45 @@
...
-                                   (float, normal_x, normal_x)
-                                   (float, normal_y, normal_y)
-                                   (float, normal_z, normal_z)
-                                   (float, rgb, rgb)
-                                   (float, weight, weight)
-                                   (unsigned int, age, age)
-                                   (std::uint32_t, directions, directions)
-                                  )
-POINT_CLOUD_REGISTER_POINT_WRAPPER (pcl::ihs::PointIHS, pcl::ihs::_PointIHS)
+POINT_CLOUD_REGISTER_POINT_STRUCT(
```
Comment (larshg) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/common_types.h:60]: This one got a bit harder to read - can we do anything about it?

Comment (SunBlack) [reply apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/common_types.h:60]: Yep, turning clang-tidy off for this lines. Some macros are a bit hard for automatic formatting ;-). We could check if newer version of Clang-Format have a way to define rules for this cases, but not sure.

**Comment by mvieth** [comment]
I manually ran the [documentation CI](https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=19353&view=logs&j=d0d530cd-2c5d-5eb2-7bb0-02dfe64fb679&t=6ab530e7-bcf3-5fe4-021f-7a3fca74087f) to see if there are any new doxygen warnings/errors, but it appears that [apps are excluded from doxygen](https://github.com/PointCloudLibrary/pcl/blob/master/doc/doxygen/doxyfile.in#L127). I wonder if that makes sense, because the apps headers are installed and available e.g. [via the apt package](https://packages.ubuntu.com/focal/amd64/libpcl-dev/filelist)

**Comment by SunBlack** [comment]
> I manually ran the [documentation CI](https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=19353&view=logs&j=d0d530cd-2c5d-5eb2-7bb0-02dfe64fb679&t=6ab530e7-bcf3-5fe4-021f-7a3fca74087f) to see if there are any new doxygen warnings/errors, but it appears that [apps are excluded from doxygen](https://github.com/PointCloudLibrary/pcl/blob/master/doc/doxygen/doxyfile.in#L127). I wonder if that makes sense, because the apps headers are installed and available e.g. [via the apt package](https://packages.ubuntu.com/focal/amd64/libpcl-dev/filelist)

I don't see even any reason to deliver the headers for the apps. Maybe it would be even better to split the PCL package into 3 parts: libpcl (only header), libpcl-dev (libpcl + header), pcl-apps (libpcl + apps). But I think this is not topic of this PR here ;-)

**Comment by mvieth** [comment]
For the parts that are included in [the apps library](https://packages.ubuntu.com/focal/amd64/libpcl-apps1.10/filelist), apparently mainly dominant_plane_segmentation and render_views_tesselated_sphere, it makes sense to deliver the headers and run doxygen. For the point_cloud_editor (and the other apps that are not in apps library) I too currently don't see a reason why the headers should be available. But it could still be good to run doxygen once after these formatting changes to see if the doxygen comments are still valid.
Regarding splitting the PCL package: the current situation is not totally different from what you described: there is a package for each module containing only the module library (e.g. libpcl-filters1.10, libpcl-apps1.10), and there is libpcl-dev containing all the headers (plus some other stuff), and also depending on all individual module library packages.

**Review by larshg — DISMISSED** [review]

**Review by mvieth — DISMISSED** [review]
LGTM, but we should discuss sometime whether it is correct that the point_cloud_editor headers are installed

**Review by larshg — APPROVED** [review]

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h [hunk 2]
```diff
@@ -40,172 +40,176 @@
...
+  using MeshPtr = pcl::ihs::MeshPtr;
+  using MeshConstPtr = pcl::ihs::MeshConstPtr;
+
+  /** \brief Constructor */
+  ICP();
+
+  /** @{ */
+  /** \brief Convergence is detected when the change of the fitness between the current
+   * and previous iteration becomes smaller than the given epsilon (set in cm^2). The
+   * fitness is the mean squared euclidean distance between corresponding points. \note
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h:74]: \note on separate line

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h [hunk 3]
```diff
@@ -40,172 +40,176 @@
...
+  void
+  setEpsilon(const float epsilon);
+
+  float
+  getEpsilon() const;
+  /** @} */
+
+  /** @{ */
+  /** \brief The registration fails if the number of iterations exceeds the maximum
+   * number of iterations. \note Must be greater than 0. Smaller values are set to 1.
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h:86]: Same here

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h [hunk 4]
```diff
@@ -40,172 +40,176 @@
...
+  void
+  setMinOverlap(const float overlap);
+
+  float
+  getMinOverlap() const;
+  /** @} */
+
+  /** @{ */
+  /** \brief The registration fails at the state of convergence if the fitness is bigger
+   * than this threshold (set in cm^2) \note Must be greater than zero.
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h:111]: And here

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h [hunk 5]
```diff
@@ -40,172 +40,176 @@
...
+
+  float
+  getMaxFitness() const;
+  /** @} */
+
+  /** @{ */
+  /** \brief Correspondences are rejected if the squared distance is above a threshold.
+   * This threshold is initialized with infinity (all correspondences are accepted in
+   * the first iteration). The threshold of the next iterations is set to the fitness of
+   * the current iteration multiplied by the factor set by this method. \note Must be
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h:124]: Here as well

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h [hunk 6]
```diff
@@ -40,172 +40,176 @@
...
+  void
+  setCorrespondenceRejectionFactor(const float factor);
+
+  float
+  getCorrespondenceRejectionFactor() const;
+  /** @} */
+
+  /** @{ */
+  /** \brief Correspondences are rejected if the angle between the normals is bigger
+   * than this threshold. Set in degrees. \note Must be between 180 degrees and 0.
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h:136]: ☝🏾

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h [hunk 7]
```diff
@@ -40,172 +40,176 @@
...
+   */
+  void
+  setMaxAngle(const float angle);
+
+  float
+  getMaxAngle() const;
+  /** @} */
+
+  /** \brief Find the transformation that aligns the data cloud (source) to the model
+   * mesh (target). \param[in] mesh_model Model mesh (target). \param[in] cloud_data
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h:147]: This time with \param

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h [hunk 8]
```diff
@@ -40,172 +40,176 @@
...
+  using CloudNormalConstPtr = CloudNormal::ConstPtr;
+
+  using KdTree = pcl::KdTree<PointNormal>;
+  using KdTreePtr = KdTree::Ptr;
+  using KdTreeConstPtr = KdTree::ConstPtr;
+
+  /** \brief Selects the model points that are pointing towards to the camera (data
+   * coordinate system = camera coordinate system). \param[in] mesh_model Input mesh.
+   * \param[in] T_inv Transformation that brings the model mesh into the camera
+   * coordinate system. \return Cloud containing the selected points (the connectivity
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h:171]: A \return here

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h [hunk 9]
```diff
@@ -40,172 +40,176 @@
...
+   * information of the mesh is currently not used during the registration).
+   */
+  CloudNormalConstPtr
+  selectModelPoints(const MeshConstPtr& mesh_model, const Eigen::Matrix4f& T_inv) const;
+
+  /** \brief Selects the valid data points. The input cloud is organized -> contains
+   * nans which are removed \param[in] cloud_data Input cloud. \return Cloud containing
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h:178]: 👆🏾

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h [hunk 10]
```diff
@@ -40,172 +40,176 @@
...
+   * the selected points.
+   */
+  CloudNormalConstPtr
+  selectDataPoints(const CloudXYZRGBNormalConstPtr& cloud_data) const;
+
+  /** \brief Finds the transformation that minimizes the point to plane distance from
+   * the source to the target cloud. The input clouds must be arranged to have one to
+   * one correspondences (point 0 in source corresponds to point 0 in target, point 1 in
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/icp.h:186]: And this blob as well

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/input_data_processing.h [hunk 11]
```diff
@@ -40,191 +40,324 @@
...
+  using CloudXYZRGBNormalConstPtr = CloudXYZRGBNormal::ConstPtr;
+
+  /** \brief Constructor */
+  InputDataProcessing();
+
+  /** \brief Apply the segmentation on the input cloud (XYZ and HSV).
+   * \param[in] cloud_in The input cloud.
+   * \param[out] cloud_out The segmented cloud.
+   * \param[out] cloud_discarded Cloud containing all points that were removed during
+   * the HSV segmentation. The points in the XYZ segmentation are NOT used! \return true
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/input_data_processing.h:77]: New lines for return and note

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/input_data_processing.h [hunk 12]
```diff
@@ -40,191 +40,324 @@
...
+  getZMax() const
+  {
+    return (z_max_);
+  }
+  /** @} */
+
+  /** @{ */
+  /** \brief Simple color segmentation in the HSV color space. Points inside of H - S -
+   * V min / max are discarded. H must be in the range 0 and 360, S and V in the range 0
+   * and 1. \note If you set values outside of the allowed range the member variables
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/input_data_processing.h:170]: Note here

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/integration.h [hunk 13]
```diff
@@ -40,183 +40,196 @@
...
+  bool
+  merge(const CloudXYZRGBNormalConstPtr& cloud_data,
+        MeshPtr& mesh_model,
+        const Eigen::Matrix4f& T) const;
+
+  /** \brief Outlier rejection. In each merge step points that have not been observed
+   * again age by one iteration. Points that are observed again get an age of 0. Once a
+   * point reaches the maximum age it is decided if the point is removed or kept in the
+   * mesh. A point is removed if it has not been observed from a minimum number of
+   * directions. \param[in,out] mesh The mesh which should be processed. \param[in]
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/integration.h:99]: These params

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/integration.h [hunk 14]
```diff
@@ -40,183 +40,196 @@
...
+   * cleanup Calls mesh.cleanup () if true.
+   */
+  void
+  age(const MeshPtr& mesh, const bool cleanup = true) const;
+
+  /** \brief Removes unfit vertices regardless of their age. Unfit vertices are those
+   * that have not been observed from enough directions. \param[in,out] mesh The which
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/integration.h:106]: Here too

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/integration.h [hunk 15]
```diff
@@ -40,183 +40,196 @@
...
+   */
+  void
+  setMaxSquaredDistance(const float squared_distance);
+  float
+  getMaxSquaredDistance() const;
+  /** @} */
+
+  /** @{ */
+  /** \brief Corresponding points are only averaged out if the angle between the normals
+   * is smaller than an angle threshold. \note Must be between 0 and 180. Values outside
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/integration.h:125]: This note

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/integration.h [hunk 16]
```diff
@@ -40,183 +40,196 @@
...
+   */
+  void
+  setMaxAngle(const float angle);
+  float
+  getMaxAngle() const;
+  /** @} */
+
+  /** @{  */
+  /** \brief Once a point reaches the maximum age it is decided if the point is removed
+   * or kept in the mesh. \note Must be greater than zero.
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/integration.h:136]: And this

File: apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/integration.h [hunk 17]
```diff
@@ -40,183 +40,196 @@
...
+   */
+  void
+  setMaxAge(const unsigned int age);
+  unsigned int
+  getMaxAge() const;
+  /** @} */
+
+  /** @{  */
+  /** \brief A point is removed if it has not been observed from a minimum number of
+   * directions. \note Must be greater than zero.
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/include/pcl/apps/in_hand_scanner/integration.h:146]: and this

**Review by kunaltyagi — CHANGES_REQUESTED** [review]
11/27 only

**Comment by mvieth** [comment]
Could you try @kunaltyagi 's doxygen_fixer script from pull request #5085 on your changes here?

**Comment by SunBlack** [comment]
> Could you try @kunaltyagi 's doxygen_fixer script from pull request #5085 on your changes here?

Tried it, but it misses some issues and also add some issue. So I did it now manually by searching for ` \`

**Review by mvieth — APPROVED** [review]

**Comment by mvieth** [comment]
@kunaltyagi Would you review this one again?

File: apps/in_hand_scanner/src/input_data_processing.cpp [hunk 18]
```diff
@@ -116,92 +111,85 @@ pcl::ihs::InputDataProcessing::segment (const CloudXYZRGBAConstPtr& cloud_in,
...
-        {
-          if (!hsv_inverted_) hsv_mask (r, c) = true;
+  for (MatrixXb::Index r = 0; r < xyz_mask.rows(); ++r) {
+    for (MatrixXb::Index c = 0; c < xyz_mask.cols(); ++c) {
+      const PointXYZRGBA& xyzrgb = (*cloud_in)[r * width + c];
+      const Normal& normal = (*cloud_normals)[r * width + c];
+
+      xyz_mask(r, c) = hsv_mask(r, c) = false;
+
+      if (!std::isnan(xyzrgb.x) && !std::isnan(normal.normal_x) && xyzrgb.x >= x_min &&
```
Comment (kunaltyagi) [inline apps/in_hand_scanner/src/input_data_processing.cpp:121]: Note to self: check for isnan on x and normal_x looks funky

**Review by kunaltyagi — APPROVED** [review]

**Pull request merged**